app/extensions/browser_extensions.rb

```diff
@@ -3,9 +3,12 @@
 require("browser")
 
 module Browser
-  # Prevent `browser` gem from saying that mobile DuckDuckGo browsers are bots
   class Base
     def bot?
+      # Ensure that test requests aren't from bots, unless test uses a known bot
+      return (bot.why? == Browser::Bot::KnownBotsMatcher) if Rails.env.test?
```

Comment:
This lets us drop -- in three places -- stubbing this method.

Reply:
I'd rather we kept the stubbing.

Reply:
Thanks for looking through this PR.
We were stubbing this method with `.any_instance.stubs`.
I'm not sure I can find a thread-safe alternative, but I will try.

Reply:
@mo-nathan:
I still cannot think of a way to stub `bot?` in the three test case files without using `any_instance.stubs`.
Any ideas?

Reply:
I'm not sure exactly what you're looking at, but I looked at the tests in `test/models/browser_extensions_test.rb` and these seems like testing code in other people's libraries.  Why do we care what their library responds with in our testing environment?  I'd be totally cool with changing the test to just `assert(Browser.new("").bot?)` and reverting the code back to the way it was.

Reply:
@mo-nathan:
Thanks again for all the time you've spent looking at this and for your comments.
I agree that we should not test other people's libraries.
Here's what I'm looking at.
- main has 3 files with this non-thread-safe stubs of `bot?`
```ruby
    # Treat Rails html requests as coming from non-robots.
    # If it's a bot, controllers often do not serve the expected content.
    # The requester looks like a bot to the `browser` gem because the User Agent
    # in the request is blank. I don't see an easy way to change that. -JDC
    Browser::Bot.any_instance.stubs(:bot?).returns(false)
```
(The three files are: test/application_system_test_case.rb, test/capybara_integration_test_case.rb, test/integration_test_case.rb.)
- I'd like to get rid of those stubs (and remove `mocha`) for the reasons you gave [here](https://github.com/MushroomObserver/mushroom-observer/pull/1664#discussion_r1350611944).  That's the point of this PR.
- If I simply remove those stubs, then (of course) I get a ton of failures and errors
```sh
vagrant@ubuntu-focal:/vagrant/mushroom-observer$ rails test:all
...
  2318/2318: [==================================] 100% Time: 00:02:44, Time: 00:02:44
Finished in 164.91208s
2322 tests, 25912 assertions, 70 failures, 19 errors, 0 skips
```
- My solution was to extend the Browser gem. 
**QUESTION: Can you suggest another solution?**
- I thought I should also unit-test the extension (as opposed to the gem _per se_).
**QUESTION: Is this a bad practice?** (We do it a lot. Seems like a good practice to me. We want to know if updates to the gems break our extensions.)
- Maybe we can discuss this at the tech meeting next Sat. I'm really jammed up this week.